lib/TPP/TileConsumerAndFuseProducers.cpp

```diff
@@ -25,108 +27,92 @@ using namespace mlir;
 
 namespace {
 
-// Check if loopRange is statically known.
-static bool isStaticRange(Range loopRange) {
-  Optional<int64_t> offsetAsInt = getConstantIntValue(loopRange.offset);
-  if (!offsetAsInt)
-    return false;
-  Optional<int64_t> sizeAsInt = getConstantIntValue(loopRange.size);
-  if (!sizeAsInt)
-    return false;
-  Optional<int64_t> strideAsInt = getConstantIntValue(loopRange.stride);
-  return static_cast<bool>(strideAsInt);
+// TODO: improve checks here.
+static bool isBlockedMatmul(linalg::GenericOp genericOp) {
+  return tpp::utils::hasMatmulBody(genericOp);
 }
 
-static int64_t getSizeRange(Range loopRange) {
-  assert(isStaticRange(loopRange));
-  return *getConstantIntValue(loopRange.size) -
-         *getConstantIntValue(loopRange.offset);
+static bool isMatmulLike(Operation *op) {
+  if (isa_and_nonnull<linalg::MatmulOp>(op))
+    return true;
+  if (linalg::GenericOp maybeMatmul = dyn_cast_or_null<linalg::GenericOp>(op))
+    return isBlockedMatmul(maybeMatmul);
+  return false;
 }
 
-static LogicalResult tileDivideIterationDomain(linalg::GenericOp linalgOp,
-                                               ArrayRef<int64_t> tiles,
-                                               OpBuilder &builder) {
-  if (linalgOp.getNumLoops() != tiles.size())
-    return failure();
-  SmallVector<Range> iterationDomain =
-      cast<TilingInterface>(linalgOp.getOperation())
-          .getIterationDomain(builder);
-  for (const auto &it : llvm::enumerate(iterationDomain)) {
-    // fine, we are not tiling along this dimension.
-    if (tiles[it.index()] == 0)
-      continue;
-    // require static loop range
-    if (!isStaticRange(it.value()))
-      return failure();
-    int64_t sizeRange = getSizeRange(it.value());
-    // fail if the tail size equals the range
-    // or is not a full tile.
-    if (tiles[it.index()] == sizeRange)
-      return failure();
-    if (sizeRange % tiles[it.index()] != 0)
-      return failure();
+static bool hasMatmulLikeProducer(linalg::LinalgOp linalgOp) {
+  for (Value operand : linalgOp->getOperands()) {
+    Operation *op = operand.getDefiningOp<linalg::LinalgOp>();
+    if (op && isMatmulLike(op))
+      return true;
   }
-  return success();
+  return false;
 }
 
-struct FuseGenericOp : public OpRewritePattern<linalg::GenericOp> {
-  FuseGenericOp(MLIRContext *context, ArrayRef<int64_t> tileSizes,
-                PatternBenefit benefit = 1)
-      : OpRewritePattern<linalg::GenericOp>(context, benefit),
-        tileSizes(tileSizes) {}
+static FailureOr<scf::SCFTilingResult>
+tileConsumer(RewriterBase &rewriter, TilingInterface consumer,
+             ArrayRef<int64_t> tileSizes) {
+  // If tileSizes is provided use them otherwise defaul to 1, 1, meaning fuse
+  // entire the two outermost loops.
+  SmallVector<int64_t> actualTileSizes = tileSizes.empty()
```

Comment:
I wonder if this should have a default value at all, or if every call should make sure to pass the tiling sizes, to make sure we don't have surprise results, for instance, if the user does pass `tileSizes` but for some buggy reason, it's empty.

I'd favour to assert if there are no sizes.

Reply:
I am not sure either. I was thinking this in the context of having a default pipeline.

Reply:
In the default pipeline we should try to adjust the sizes dynamically. But in the context of this pass on its own, it might be good QoL to have sth that works out of the box.

Reply:
The way it's done, we can introduce bugs by passing an empty container. So, if we want a default value, we need to make it default on the function declaration (`arg = {1,1}`), if not, we assert if it's empty.

I think we should start asserting, and only add a default when we're certain of the default we actually want. It's not clear to me yet that this is the default.

Reply:
Ok, so we prefer passing tile sizes to the pass and avoid default for now, right? {1, 1} works relatively well for matmul + elemenwise because the output of the matmul is 2d, so you end up with a reduction + scalar and 2 parallel loops. Or if the matmul is packed with a 2d element wise and a brgemm op.

Reply:
Not tile sizes to the pass, to the function that calls this one. Ie. just move the default out of this "generic" function, so that its usage is clear and safe.